svc/pkg/mm/worker/redis-scripts/lobby_find.lua

```diff
@@ -64,16 +70,36 @@ elseif query.kind.lobby_group ~= nil then
 		-- instead of `max_players_party` because `max_players_party` is
 		-- only relevant when joining a lobby with the exact lobby ID (i.e.
 		-- where `max_players_direct` would normally be used).
-		local lobby = redis.call('ZRANGEBYSCORE', key_available_spots, table.getn(query.players), '+inf', 'WITHSCORES',
+		local lobby = redis.call('ZRANGEBYSCORE', key_available_spots, #query.players, '+inf', 'WITHSCORES',
 			'LIMIT', '0', '1')
-		if table.getn(lobby) > 0 then
+		if #lobby > 0 then
 			local lobby_id = lobby[1]
 			local available_spots = tonumber(lobby[2])
+			local correct_tags = 0
+
+			-- Verify tags
+			if #lobby_tag_names > 0 then
+				-- Fetch specified lobby tags
+				local key_lobby_tags = '{global}:mm:lobby:' .. lobby_id .. ':tags'
+				local lobby_tags = redis.call('HMGET', key_lobby_tags, unpack(lobby_tag_names))
+				
+				-- Check that all tags are correct
+				for tag_idx, tag in ipairs(lobby_tags) do
+					if query.tags[lobby_tag_names[tag_idx]] ~= tag then
+						break
+					end
+
+					correct_tags = correct_tags + 1
+				end
+			end
 
-			-- Set as best lobby if no lobby selected or lobby is more full
-			if best_lobby_id == nil or available_spots < best_available_spots then
-				best_lobby_id = lobby_id
-				best_available_spots = available_spots
+			-- Check if the correct number of tags have been parsed
+			if correct_tags == #lobby_tag_names then
+				-- Set as best lobby if no lobby selected or lobby is more full
+				if best_lobby_id == nil or available_spots < best_available_spots then
+					best_lobby_id = lobby_id
+					best_available_spots = available_spots
```

Comment:
I think this is missing a big piece: if the lobby doesn't match the given tags, it will return immediately as no lobby found. Instead, it should iterate to the _next_ lobby and keep checking them.

e.g. If no tags are provided keep the current `ZRANGEBYSCORE ... LIMIT 0 1`. If tags are provided, then use `ZRANGEBYSCORE ... LIMIT 0 128` to check the top 128 lobbies. If there's more than 128 lobbies, then the system will break. We'll fix that before it's a problem.

Reply:
The current behavior already iterates to the next lobby if the tags don't match. I think you might be confusing the `break` statement with breaking the top most loop; it only breaks the inner loop iterating over tags.

I'm not sure what the second part of your message is describing. Are you saying it should only check at most 128 lobbies when tags are present?

Reply:
I don't see the loop you're referring to that keeps searching the same available spots sorted key for a matching lobby. Unless I'm misreading the code, it only checks one lobby in each `available_spots_key` and if it doesn't match the first lobby, goes to the next available spots key.

The `ZRANGEBYSCORE` recommendation is to let you get all of the lobby IDs in that sorted set and iterate over the lobby IDs until you find one that matches the given tags.

Reply:
https://github.com/rivet-gg/rivet/pull/197/commits/4d08c9ca1fe72741435e436e5e7aa8426c25943d